src/common/input/parameter_keymap.cpp

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright © 2021 Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License version 2 or 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Authored By: William Wold <william.wold@canoncical.com>
+ */
+
+#include "mir/input/parameter_keymap.h"
+
+#include <stdexcept>
+#include <boost/throw_exception.hpp>
+#include <xkbcommon/xkbcommon.h>
+
+namespace mi = mir::input;
+
+auto mi::ParameterKeymap::operator==(Keymap const& other) const -> bool
+{
+    auto const params = dynamic_cast<ParameterKeymap const*>(&other);
+    return params &&
```

Comment:
OK, so this is implementing "equal iff rhs is a similar type (the same type, or derived from the same type) and has the same data values"? That isn't an equivalence relationship as it isn't symmetric.

Using `operator==()` is best kept for equivalence relationships.

Reply:
For example, if we rename the existing implementation as `same(Keymap const& other)` you could have:
```C++
inline auto operator==(Keymap const& lhs, Keymap const& rhs)
{
    return lhs.same(rhs) && rhs.same(lhs);
}
```

Reply:
Done